xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+        }
+        return result;
+    }
+
+    /**
+     * Create and/or update objects in a document given a list of HTTP parameters of the form
+     * '<spacename>.<classname>_<number>_<propertyname>'. If the object already exists, the field is replace by the
+     * given value. If the object doesn't exist in the document, it is created then the property <propertyname> is
+     * initialized with the given value. An object is only created if the given '<number>' is 'one-more' than the
+     * existing number of objects. For example, if the document already has 2 objects of type 'Space.Class', then it
+     * will create a new object only with 'Space.Class_2_prop=something'. Every other parameter like
+     * 'Space.Class_42_prop=foobar' for example, will be ignore.
+     * 
+     * @param eform Form information that contains all the query parameters
+     * @param context
+     * @throws XWikiException
+     * @since 7.0RC1
+     */
+    public void readObjectsFromFormUpdateOrCreate(EditForm eform, XWikiContext context) throws XWikiException
+    {
+        Map<String, SortedMap<Integer, Map<String, String[]>>> fromRequest =
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<String, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest.entrySet()) {
+            DocumentReference requestClassReference =
+                getCurrentDocumentReferenceResolver().resolve(requestClassEntries.getKey());
+            SortedMap<Integer, Map<String, String[]>> requestObjectMap = requestClassEntries.getValue();
+            List<BaseObject> newObjects = new ArrayList<>();
+            for (Entry<Integer, Map<String, String[]>> requestObjectEntry : requestObjectMap.entrySet()) {
+                Integer requestObjectNumber = requestObjectEntry.getKey();
+                Map<String, String[]> requestObjectPropertyMap = requestObjectEntry.getValue();
+                BaseObject oldObject = getXObject(requestClassReference, requestObjectNumber);
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
+                    if (requestObjectPropertyMap != null
+                        && requestObjectNumber == getXObjectSize(requestClassReference)) {
+                        oldObject = newXObject(requestClassReference, context);
+                    } else {
+                        break;
+                    }
+                }
+                BaseClass baseClass = oldObject.getXClass(context);
+                BaseObject newObject = (BaseObject) baseClass.fromMap(requestObjectPropertyMap, oldObject);
+                newObject.setNumber(oldObject.getNumber());
+                newObject.setGuid(oldObject.getGuid());
+                newObject.setOwnerDocument(this);
+                newObjects.add(newObject);
+            }
+            getXObjects().put(requestClassReference, newObjects);
```

Comment:
[`java.util.Map.put()`](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#put%28K,%20V%29), which is the object returned by `getXObjects()` method, doesn't replace content for different keys.  In this case, other classes contains other keys so they should not be destroyed.